File: tests/passwords/test_simple_7.py
# Copyright (c) 2025-2026 Arista Networks, Inc.
# Use of this source code is governed by the Apache License 2.0
# that can be found in the LICENSE file.

from contextlib import AbstractContextManager
from contextlib import nullcontext as does_not_raise

import pytest

from pyavd_utils.passwords import (
    PyAVDUtilsPasswordError,
    Simple7DataTooShortError,
    Simple7EmptyPasswordError,
    Simple7Error,
    Simple7InvalidHexEncodingError,
    Simple7InvalidSaltFormatError,
    Simple7InvalidSaltValueError,
    simple_7_decrypt,
    simple_7_encrypt,
)


def test_simple_7_error_hierarchy() -> None:
    """Test that Type-7 errors inherit from the passwords base error."""
    assert issubclass(Simple7Error, PyAVDUtilsPasswordError)
    assert issubclass(Simple7InvalidSaltValueError, Simple7Error)
    assert issubclass(Simple7EmptyPasswordError, Simple7Error)


SIMPLE_7_ENCRYPT_TEST_DATA = [
    pytest.param(
        "foo",
        1,
        "0115090B",
        does_not_raise(),
        id="Valid encryption with salt 1",
    ),
    pytest.param(
        "foo",
        6,
        "0600002E",
        does_not_raise(),
        id="Valid encryption with salt 6",
    ),
    pytest.param(
        "foo",
        9,
        "094A4106",
        does_not_raise(),
        id="Valid encryption with salt 9",
    ),
    pytest.param(
        "foo",
        15,
        "15140403",
        does_not_raise(),
        id="Valid encryption with salt 15",
    ),
    pytest.param(
        "test_password",
        16,
        "",
        pytest.raises(Simple7InvalidSaltValueError, match="Salt must be in the range 0-15, got 16"),
        id="Invalid salt value (16)",
    ),
    pytest.param(
        "test_password",
        99,
        "",
        pytest.raises(Simple7InvalidSaltValueError, match="Salt must be in the range 0-15, got 99"),
        id="Invalid salt value (99)",
    ),
    pytest.param(
        "",
        5,
        "",
        pytest.raises(Simple7EmptyPasswordError, match="Password must not be empty"),
        id="Empty password",
    ),
]


@pytest.mark.parametrize(("data", "salt", "expected_encrypted", "expected_raise"), SIMPLE_7_ENCRYPT_TEST_DATA)
def test_simple_7_encrypt(data: str, salt: int, expected_encrypted: str, expected_raise: AbstractContextManager[None]) -> None:
    """Test simple_7_encrypt success and various failures."""
    with expected_raise:
        assert simple_7_encrypt(data, salt) == expected_encrypted


SIMPLE_7_DECRYPT_TEST_DATA = [
    pytest.param(
        "0115090B",
        "foo",
        does_not_raise(),
        id="Valid decryption with salt 1",
    ),
    pytest.param(
        "0600002E",
        "foo",
        does_not_raise(),
        id="Valid decryption with salt 6",
    ),
    pytest.param(
        "094A4106",
        "foo",
        does_not_raise(),
        id="Valid decryption with salt 9",
    ),
    pytest.param(
        "15140403",
        "foo",
        does_not_raise(),
        id="Valid decryption with salt 15",
    ),
    pytest.param(
        "",
        "",
        pytest.raises(Simple7DataTooShortError, match="Encrypted data too short"),
        id="Data too short (empty)",
    ),
    pytest.param(
        "0",
        "",
        pytest.raises(Simple7DataTooShortError, match="Encrypted data too short"),
        id="Data too short (1 char)",
    ),
    pytest.param(
        "01GGGG",
        "",
        pytest.raises(Simple7InvalidHexEncodingError, match="Invalid hex encoding"),
        id="Invalid hex encoding",
    ),
    pytest.param(
        "XX1234",
        "",
        pytest.raises(Simple7InvalidSaltFormatError, match="Invalid salt format"),
        id="Invalid salt format",
    ),
    pytest.param(
        "161234",
        "",
        pytest.raises(Simple7InvalidSaltValueError, match="Salt must be in the range 0-15, got 16"),
        id="Salt out of range (16)",
    ),
    pytest.param(
        "991234",
        "",
        pytest.raises(Simple7InvalidSaltValueError, match="Salt must be in the range 0-15, got 99"),
        id="Salt out of range (99)",
    ),
]


@pytest.mark.parametrize(("encrypted_data", "expected_plain", "expected_raise"), SIMPLE_7_DECRYPT_TEST_DATA)
def test_simple_7_decrypt(encrypted_data: str, expected_plain: str, expected_raise: AbstractContextManager[None]) -> None:
    """Test simple_7_decrypt success and various failures."""
    with expected_raise:
        assert simple_7_decrypt(encrypted_data) == expected_plain


SIMPLE_7_ROUNDTRIP_TEST_DATA = [
    pytest.param(
        "test_password",
        None,
        id="Roundtrip with random salt",
    ),
    pytest.param(
        "arista",
        5,
        id="Roundtrip with salt 5",
    ),
    pytest.param(
        "another_password",
        0,
        id="Roundtrip with salt 0",
    ),
]


@pytest.mark.parametrize(("password", "salt"), SIMPLE_7_ROUNDTRIP_TEST_DATA)
def test_simple_7_roundtrip(password: str, salt: int | None) -> None:
    """Test simple_7 encrypt/decrypt roundtrip."""
    encrypted = simple_7_encrypt(password, salt)
    decrypted = simple_7_decrypt(encrypted)
    assert decrypted == password

File: rust/pypasswords/src/tests/test_simple_7.rs
// Copyright (c) 2025-2026 Arista Networks, Inc.
// Use of this source code is governed by the Apache License 2.0
// that can be found in the LICENSE file.

use super::*;
use crate::errors::Simple7EncryptPyError;

#[test]
fn simple_7_encrypt_decrypt_roundtrip() {
    with_passwords_module(|_py, module| {
        let password = "test_password";
        let salt = 5_u8;

        let encrypted: String = module
            .call_method1("simple_7_encrypt", (password, Some(salt)))
            .unwrap()
            .extract()
            .unwrap();

        let decrypted: String = module
            .call_method1("simple_7_decrypt", (encrypted,))
            .unwrap()
            .extract()
            .unwrap();

        assert_eq!(decrypted, password);
    });
}

#[test]
fn simple_7_encrypt_with_random_salt() {
    with_passwords_module(|py, module| {
        let password = "test_password";

        // Call with None for salt
        let encrypted: String = module
            .call_method1("simple_7_encrypt", (password, py.None()))
            .unwrap()
            .extract()
            .unwrap();

        // Should be able to decrypt it
        let decrypted: String = module
            .call_method1("simple_7_decrypt", (encrypted,))
            .unwrap()
            .extract()
            .unwrap();

        assert_eq!(decrypted, password);
    });
}

#[test]
fn simple_7_encrypt_empty_password_err() {
    with_passwords_module(|py, module| {
        let err = module
            .call_method1("simple_7_encrypt", ("", Some(5_u8)))
            .unwrap_err();

        assert!(err.is_instance_of::<passwords::Simple7EmptyPasswordError>(py));
        assert!(err.is_instance_of::<passwords::PasswordError>(py));
        assert_eq!(err.value(py).to_string(), "Password must not be empty.");
    });
}

#[test]
fn simple_7_encrypt_invalid_salt_err() {
    with_passwords_module(|py, module| {
        let password = "test_password";
        let invalid_salt = 16_u8; // Out of range (0-15)

        let err = module
            .call_method1("simple_7_encrypt", (password, Some(invalid_salt)))
            .unwrap_err();

        assert!(err.is_instance_of::<passwords::Simple7InvalidSaltValueError>(py));
        assert!(err.is_instance_of::<passwords::PasswordError>(py));
        assert_eq!(
            err.value(py).to_string(),
            "Salt must be in the range 0-15, got 16."
        );
    });
}

#[test]
fn simple_7_decrypt_data_too_short_err() {
    with_passwords_module(|py, module| {
        let err = module.call_method1("simple_7_decrypt", ("0",)).unwrap_err();

        assert!(err.is_instance_of::<passwords::Simple7DataTooShortError>(py));
        assert_eq!(
            err.value(py).to_string(),
            "Encrypted data too short (minimum 2 characters required for salt)."
        );
    });
}

#[test]
fn simple_7_decrypt_invalid_hex_err() {
    with_passwords_module(|py, module| {
        let err = module
            .call_method1("simple_7_decrypt", ("01GGGG",))
            .unwrap_err();

        assert!(err.is_instance_of::<passwords::Simple7InvalidHexEncodingError>(py));
        assert!(err.value(py).to_string().contains("Invalid hex encoding"));
    });
}

#[test]
fn simple_7_decrypt_invalid_salt_format_err() {
    with_passwords_module(|py, module| {
        let err = module
            .call_method1("simple_7_decrypt", ("XX1234",))
            .unwrap_err();

        assert!(err.is_instance_of::<passwords::Simple7InvalidSaltFormatError>(py));
        assert!(err.value(py).to_string().contains("Invalid salt format"));
    });
}

#[test]
fn simple_7_decrypt_salt_out_of_range_err() {
    with_passwords_module(|py, module| {
        let err = module
            .call_method1("simple_7_decrypt", ("161234",))
            .unwrap_err();

        assert!(err.is_instance_of::<passwords::Simple7InvalidSaltValueError>(py));
        assert_eq!(
            err.value(py).to_string(),
            "Salt must be in the range 0-15, got 16."
        );
    });
}

#[test]
fn simple_7_decrypt_invalid_utf8_err() {
    with_passwords_module(|py, module| {
        // Salt 0 uses seed byte 0x64. Encrypted byte 0x9b decrypts to 0xff,
        // which is not a valid single-byte UTF-8 sequence.
        let err = module
            .call_method1("simple_7_decrypt", ("009B",))
            .unwrap_err();

        assert!(err.is_instance_of::<passwords::Simple7InvalidUtf8Error>(py));
        assert!(err.value(py).to_string().contains("not valid UTF-8"));
    });
}

#[test]
fn simple_7_random_source_unavailable_maps_to_specific_pyerr() {
    with_passwords_module(|py, _module| {
        let err = pyo3::PyErr::from(Simple7EncryptPyError::from(
            ::passwords::Simple7Error::RandomSourceUnavailable(getrandom::Error::UNSUPPORTED),
        ));

        assert!(err.is_instance_of::<passwords::Simple7RandomSourceUnavailableError>(py));
        assert!(err.is_instance_of::<passwords::PasswordError>(py));
        assert_eq!(
            err.value(py).to_string(),
            "Failed to obtain random salt from the operating system."
        );
    });
}

#[test]
fn simple_7_known_values() {
    with_passwords_module(|_py, module| {
        // Test known encryption values
        let test_cases: [(u8, &str, &str); 4] = [
            (1, "foo", "0115090B"),
            (6, "foo", "0600002E"),
            (9, "foo", "094A4106"),
            (15, "foo", "15140403"),
        ];

        for (salt, password, expected_encrypted) in test_cases {
            let encrypted: String = module
                .call_method1("simple_7_encrypt", (password, Some(salt)))
                .unwrap()
                .extract()
                .unwrap();
            assert_eq!(encrypted, expected_encrypted);

            let decrypted: String = module
                .call_method1("simple_7_decrypt", (expected_encrypted,))
                .unwrap()
                .extract()
                .unwrap();
            assert_eq!(decrypted, password);
        }
    });
}
